### ISSUE TITLE
Mismatch of VM data and map data related to overgeneration

### ISSUE THREAD

**Issue by kromka-chleba** [minetest/minetest#14526]
### Minetest version

```true
commit: 6a7a613741fe32fa14897a4f780d7382105a103c

Minetest 5.9.0-dev (Linux)
Using Irrlicht 1.9.0mt15
Using LuaJIT 2.1.0-beta3
Built by GCC 11.3
Running on Linux/6.7.10 x86_64
BUILD_TYPE=RelWithDebInfo
RUN_IN_PLACE=0
USE_CURL=1
USE_GETTEXT=1
USE_SOUND=1
STATIC_SHAREDIR="/gnu/store/sphglsqv01pss5m5dx4p3abc89rl780p-minetest-5.9.0-dev-1.6a7a613/share/minetest"
STATIC_LOCALEDIR="/gnu/store/sphglsqv01pss5m5dx4p3abc89rl780p-minetest-5.9.0-dev-1.6a7a613/share/locale"
```

### Operating system and version

Guix System, commit: fa98837b1d9586b9ca7bf77dae22c3bde0fc59b5

### CPU model

AMD Ryzen 7 2700 (16) @ 3.200GHz

### GPU model

AMD ATI Radeon RX 6600/6600 XT/6600M

### Summary

I'm writing a road generating system using Voxel Manipulators and I noticed weird holes in the roads.
I believe that the cause is that data obtained with vm:get_data() does not match map data.

A picture is worth more than thousand words so:

gray lines - mapchunk borders
red (brick) lines - borders where mapgen overgenerates stuff

![sus](https://github.com/minetest/minetest/assets/65868699/9d25f50e-d279-4b51-b2c9-7955dd6d698e)

![sus3](https://github.com/minetest/minetest/assets/65868699/6de2c64b-1297-4275-9488-d1bfbde87cc8)

The code only places the lines if there's no asphalt there, yet clearly you can see holes where asphalt should be but is absent. The reason - in VM data grass nodes were replaced with asphalt so bricks and concrete were not placed, yet mapgen failed to actually write that data to map.

The code that does it is here:
https://codeberg.org/lord_of_the_dumpster/perfect_city/src/commit/72d02cdd02a399a32842058fef024b0469b1c991/mods/pcity_mapgen/roads_mapgen.lua#L82
https://codeberg.org/lord_of_the_dumpster/perfect_city/src/commit/72d02cdd02a399a32842058fef024b0469b1c991/mods/pcity_mapgen/mapgen.lua#L73

Mapgen is started with the new stuff: `minetest.register_mapgen_script(mod_path.."/mapgen.lua")`

### Steps to reproduce

1. Download my game: 
https://codeberg.org/lord_of_the_dumpster/perfect_city
`git clone https://codeberg.org/lord_of_the_dumpster/perfect_city.git`
2. Switch to the wip-roadgen branch: `git checkout wip-roadgen`
3. Start a new world
4. Turn on privs and start flying around
5. Find a road and follow it, you should be able to see not that uncommon bugs of that kind

**Comment by SmallJoker** [comment]
Does this also happen when you only write to the VManip data between `minp` and `maxp` (`minetest.register_on_generated` callback) ? The emerged area is usually larger to allow spawning of overhanging structures (e.g. trees), thus the mapgen might get confused if there's too much written to it. That would also help to narrow down the issue.

**Comment by kromka-chleba** [comment]
>Does this also happen when you only write to the VManip data between minp and maxp 

The holes still happen but the issue with disappearing lines appears to be gone.
Should I update the branch with new code? It's quite a mess though.

![Screenshot_2024-04-06_22-13-14](https://github.com/minetest/minetest/assets/65868699/2004e6c2-f260-44c5-af2b-abc8af33ac95)
![Screenshot_2024-04-06_22-10-35](https://github.com/minetest/minetest/assets/65868699/68e583ef-dab3-4226-a2af-1da98fc617d9)
![Screenshot_2024-04-06_22-09-38](https://github.com/minetest/minetest/assets/65868699/c28ae1d6-e301-4be1-8e12-b2baebb36048)

**Comment by kromka-chleba** [comment]
This one is probably the most brutal example, doesn't look like a coincidence:

![Screenshot_2024-04-06_22-40-38](https://github.com/minetest/minetest/assets/65868699/b3f7e877-7779-4f95-9094-0cc4a2349c44)

**Comment by savilli** [comment]
I think the emerged area is not only one more than block size, but the actual mapgen also fills the larger area. In this case, mapgen running on a block fills the borders of its neighbors with grass and overrides the roads. That's why you can't fill the area only between minp and maxp - you also have to put back roads on the neighbor blocks.

**Comment by kromka-chleba** [comment]
>That's why you can't fill the area only between minp and maxp - you also have to put back roads on the neighbor blocks.

The bug is here with or without overgeneration. The only explanation I see viable is mapgen failing to fully write vm data to map. Though I need a confirmation from a dev familiar with mapgen code.

**Comment by sfan5** [comment]
As already pointed out you should only write to the area between `minp` and `maxp`.

A few details are unclear to me however:
* who generates the grass?
* are you sure those position calculations in `pcmg.generate_road` are correct? Using the `VoxelArea` helper would be much better.

> I believe that the cause is that data obtained with vm:get_data() does not match map data.

Did you attempt to confirm this theory?
You have to move your code back to the normal environment, but then you could simply compare the results from `minetest.get_node` and the vmanip.

**Comment by kromka-chleba** [comment]
>As already pointed out you should only write to the area between minp and maxp.

I'm already doing this since this comment: https://github.com/minetest/minetest/issues/14526#issuecomment-2041183676

>who generates the grass?

Flat mapgen.

>are you sure those position calculations in pcmg.generate_road are correct? Using the VoxelArea helper would be much better.

Yes, tested it many times. This also happens with VoxelArea: https://codeberg.org/lord_of_the_dumpster/perfect_city/src/commit/38cd37fa038e79e568046f993b40acf561465b02/mods/pcity_mapgen/roads_mapgen.lua#L108
Also I developed these weird calculations because they're slightly faster than VA.

>Did you attempt to confirm this theory?

Not yet, but I can try.

>You have to move your code back to the normal environment, but then you could simply compare the results from minetest.get_node and the vmanip.

Ok, will try this.

**Comment by kromka-chleba** [comment]
>Did you attempt to confirm this theory?

Okay, turns out my theory is wrong which leaves me clueless.
I made a table that contains positions of all generated asphalt nodes that corresponds to VM data and made an item that checks if a clicked node is in the table. Turns out the gaps are not in VM data.
It appears to me that some of the above bugs were due to me overgenerating chunks, but what's left now appears to be a bug on my side.
With the current code the breaks only start at mapchunk borders and never on the red lines where stuff is overgenerated.

**Comment by sfan5** [comment]
> EDIT: Looks like I'm a clown and the bug is my fault, sorry for your time

Now you gotta tell us what the bug was

**Comment by appgurueu** [comment]
> EDIT: Looks like I'm a clown and the bug is my fault, sorry for your time

Better than being left with a vague feeling of "Minetest (or my code) is broken", and now you can fix it :)

There will always be false positives, the best you can do is try to minimize the chances of reporting one (for example by trying to extract a minimal reproducible example from your code before reporting; this is often a step of debugging anyways).

**Comment by kromka-chleba** [comment]
>Now you gotta tell us what the bug was :)

@sfan5
The bug was caused by me not applying the mapchunk offset (xyz -32) when I converted my "citychunk" units (10 mapchunks = 800 nodes) to normal grid coordinates.
I still see some bugs that occur only in the overgeneration area though. I'm going to make a minimal test to replicate it and will reopen the issue if/when I make sure it's not something in my code again.
